[Merged by Bors] - feat(data/nat/factorial) : descending factorial

ericrbg opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

Essentially `pochhammer`, but for nat. Used in #7363 and #7526.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ericrbg committed:
Move desc_fac out

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +143,37 @@ begin
+  induction k with t ht, simp!, rw desc_fac_succ, rw desc_fac_succ,
+  have : (n + 1) * ((n.succ + t + 1) * desc_fac n.succ t)
+       = (n.succ + t + 1) * ((n + 1) * desc_fac n.succ t), by ac_refl,
+  rw this, rw ht, repeat {rw nat.succ_eq_add_one}, ac_refl
```
```suggestion
rw [this, ht, nat.succ_add],
```

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +143,37 @@ begin
+
+lemma succ_desc_fac {n k : ℕ} : (n + 1) * desc_fac n.succ k = (n + k + 1) * desc_fac n k :=
+begin
+  induction k with t ht, simp!, rw desc_fac_succ, rw desc_fac_succ,
```
The `simp` disposes of the base case on its own, right? Meaning the `rw desc_fac_succ`s are applied to the second goal? If so they should go on a separate line (or you can combine them with the `rw`s on the last line).

ericrbg replied:
yep, for sure!

ericrbg committed:
golf

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

ericrbg committed:
golf right

jcommelin commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +143,37 @@ begin
+section desc_fac
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation.
+This is closely related to `ring_theory.polynomial.pochhammer`, but much less general. -/
```
Is there an obstruction to making this relation explicit? Please add the specialization glue.

ericrbg replied:
doing it! is it worth stripping out some of the explicit results about ℕ in `pochhammer` and making them about `desc_fac` (e.g. `factorial_mul_pochhammer'`)? cc @semorrison

ericrbg replied:
argh, there's also a little work to do in `bernstein` too... I think `iterate_derivative_at_0_aux₁` basically comes for free, though, I used it in `card_embedding` so I'll make this API more expansive

eric-wieser commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +143,37 @@ begin
+end
+
+/-- Prove that `desc_fac` is what it is promised to be. Stated divison-less for ease. -/
+theorem eval_desc_fac (n : ℕ) : ∀ k : ℕ, (n + k)! = n! * desc_fac n k
```
I think it's worth stating the version with division too, just to make it obvious the definition is really what you say it is.

ericrbg committed:
Merge branch 'master' into desc_fac

ericrbg committed:
Please let me have both CI runs!

ericrbg committed:
try adding @[simp]

ericrbg committed:
not make an absolute disaster out of everything

ericrbg committed:
this is a nightmare; think this is the last one

ericrbg committed:
Try removing @[simp]

ericrbg committed:
should finally be good

ericrbg committed:
it should not be a `simp` lemma

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,51 @@ begin
+theorem eval_desc_fac (n : ℕ) : ∀ k, n! * desc_fac n k = (n + k)!
+| 0 := by rw [desc_fac, add_zero, mul_one]
+| (k + 1) := eq.symm $
+             calc (n + (k + 1))! = (n + k + 1) * (n + k)! : by rw [←add_assoc, factorial]
```
Is there a reason why you use symmetry and don't just write the `calc` bloc in the other direction?

ericrbg replied:
it would go over the 100char limit otherwise, which is fixable but ugly; not sure what's preferred

benjamindavidson replied:
Hmm, let me take a closer look.

benjamindavidson replied:
I tried the following, which fits within the 100char limit.
```
| (k + 1) := calc n! * desc_fac n (k + 1) = n! * ((n + k + 1) * desc_fac n k) : by rw desc_fac_succ
                                      ... = (n + k + 1) * (n! * desc_fac n k) : by ac_refl
                                      ... = (n + k + 1) * (n + k)! : by rw eval_desc_fac k
                                      ... = (n + (k + 1))! : by rw [←add_assoc, factorial]
```

benjamindavidson replied:
Actually, you can just do
```
| (k + 1) := by rw [desc_fac_succ, mul_left_comm, eval_desc_fac, ← add_assoc, factorial]
```

ericrbg replied:
woah, i've been looking for these lemmas that make assoc+comm work well; thought i'd have to do a lot of nonsense. that's amazing, thanks! i'm gonna neaten up a lot of places where I used `ac_refl` for this

benjamindavidson replied:
Glad to be of help! I prefer these as well, though it is nice to let the automation do the dirty work for you sometimes :)

ericrbg committed:
improve proof

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,50 @@ begin
+  have : (n + 1) * ((n.succ + k + 1) * desc_fac n.succ k)
+       = (n.succ + k + 1) * ((n + 1) * desc_fac n.succ k), by ac_refl,
+  rw [desc_fac, this, succ_desc_fac k, desc_fac, succ_eq_add_one], ac_refl
+end
```
Similarly, 
```suggestion
| (k + 1) := by rw [desc_fac, mul_left_comm, succ_desc_fac, desc_fac, succ_add, ← add_assoc]
```
(note how `.succ` is defeq to ` + 1`)

ericrbg replied:
~~I was thinking of stopping fighting the order it wants to be and make it `n+1+k`, even though I'd never write that in paper; does that seem reasonable? This appears a little in my rewrite of `pochhammer too~~ scratch that

ericrbg commented on src/ring_theory/polynomial/pochhammer.lean:
```diff
@@ -102,6 +103,14 @@ begin
+| (t + 1) := begin
+  rw [pochhammer_succ_right, eval_mul, pochhammer_nat_eq_desc_fac t],
+  suffices : n.desc_fac t * (n + 1 + t) = n.desc_fac (t + 1), by simpa,
+  rw nat.desc_fac_succ, ac_refl
```
```suggestion
  rw [nat.desc_fac_succ, add_right_comm, mul_comm]
```

ericrbg commented on src/ring_theory/polynomial/pochhammer.lean:
```diff
@@ -123,49 +132,27 @@ section factorial
+  exact or.inl h,
+  exact or.inr rfl,
 end
+| (k + 1) := by norm_cast; repeat { rw [pochhammer_nat_eq_desc_fac] }; rw nat.succ_desc_fac; ac_refl
```
```suggestion
| (k + 1) := by norm_cast; simp only [pochhammer_nat_eq_desc_fac, nat.succ_desc_fac, add_right_comm]
```

ericrbg commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,50 @@ begin
+| (k + 1) := calc n! * desc_fac n (k + 1) = n! * ((n + k + 1) * desc_fac n k) : by rw desc_fac_succ
+                                      ... = (n + k + 1) * (n! * desc_fac n k) : by ac_refl
+                                      ... = (n + k + 1) * (n + k)! : by rw eval_desc_fac k
+                                      ... = (n + (k + 1))! : by rw [←add_assoc, factorial]
```
```suggestion
| (k + 1) := by rw [desc_fac_succ, mul_left_comm, eval_desc_fac, ← add_assoc, factorial]
```

ericrbg committed:
`ac_refl` begone

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,42 @@ begin
+| (k + 1) := by rw [desc_fac, mul_left_comm, succ_desc_fac, desc_fac, succ_add, ← add_assoc]
+
+/-- Prove that `desc_fac` is what it is promised to be. Stated divison-less for ease. -/
+theorem eval_desc_fac (n : ℕ) : ∀ k, n! * desc_fac n k = (n + k)!
```
Can you please improve this docstring so that it is independent of other docstrings? Meaning, the way it is now, I would need to refer to the docstring provided for the definition of `desc_fac` in order to see what is meant by "what it is promised to be," but docstrings should be context-free.

ericrbg committed:
docstring updates

jcommelin commented on src/data/nat/factorial.lean:
```diff
@@ -170,12 +170,12 @@ lemma succ_desc_fac (n : ℕ) : ∀ k, (n + 1) * desc_fac n.succ k = (n + k + 1)
 | (k + 1) := by rw [desc_fac, mul_left_comm, succ_desc_fac, desc_fac, succ_add, ← add_assoc]
 
-/-- Prove that `desc_fac` is what it is promised to be. Stated divison-less for ease. -/
+/-- `desc_fac n k = (n + k)! / n!`. Stated divison-less for simpicity. -/
```
```suggestion
/-- `desc_fac n k = (n + k)! / n!`. However, this lemma states a reformulation to avoid `nat`-division.
See `eval_desc_fac'` if you really need the version that uses `nat`-division. -/
```

ericrbg committed:
Update src/data/nat/factorial.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

ericrbg committed:
avoid 100-char linter

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+section desc_fac
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation.
+This is closely related to `ring_theory.polynomial.pochhammer`, but much less general. -/
```
```suggestion
This is closely related to `pochhammer`, but much less general. -/
```

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+| (k + 1) := by rw [desc_fac, mul_left_comm, succ_desc_fac, desc_fac, succ_add, ← add_assoc]
+
+/-- `desc_fac n k = (n + k)! / n!`. However, this lemma states a reformulation to avoid ℕ-division.
+See `eval_desc_fac'` if you really need the version that uses ℕ-division. -/
```
```suggestion
See `nat.eval_desc_fac'` if you really need the version that uses ℕ-division. -/
```

ericrbg replied:
why do we have to put `nat.` here even though we're in `nat`, whilst we can get rid of the `ring_theory.polynomial` above? is the docgen thing that clever?

eric-wieser replied:
I think you're confusing filenames with namespaces. Doc-gen is not clever enough to infer namespaces, but lean doesn't care about filenames.

benjamindavidson commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+| 0 := by rw [desc_fac, add_zero, mul_one]
+| (k + 1) := by rw [desc_fac_succ, mul_left_comm, eval_desc_fac, ← add_assoc, factorial]
+
+/-- Avoid in favour of `eval_desc_fac` if you can. ℕ-division isn't worth it. -/
```
```suggestion
/-- Avoid in favour of `nat.eval_desc_fac` if you can. ℕ-division isn't worth it. -/
```

eric-wieser commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+| (k + 1) := by rw [desc_fac_succ, mul_left_comm, eval_desc_fac, ← add_assoc, factorial]
+
+/-- Avoid in favour of `eval_desc_fac` if you can. ℕ-division isn't worth it. -/
+lemma eval_desc_fac' (n k : ℕ) : desc_fac n k = (n + k)! / n! :=
```
I think a better name would be
```suggestion
lemma desc_fac_eq_div (n k : ℕ) : desc_fac n k = (n + k)! / n! :=
```

eric-wieser commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+
+/-- `desc_fac n k = (n + k)! / n!`. However, this lemma states a reformulation to avoid ℕ-division.
+See `eval_desc_fac'` if you really need the version that uses ℕ-division. -/
+theorem eval_desc_fac (n : ℕ) : ∀ k, n! * desc_fac n k = (n + k)!
```
This name seems more representative of the statement
```suggestion
theorem factorial_mul_desc_fac (n : ℕ) : ∀ k, n! * desc_fac n k = (n + k)!
```

eric-wieser commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+
+section desc_fac
+
+/-- desc_fac n k = (n + k)! / n!, but implemented in a recursive way for calculation.
```
```suggestion
/-- `desc_fac n k = (n + k)! / n!` (as shown in `nat.desc_fac_eq_div`), but implemented in a recursive way for calculation.
```

ericrbg committed:
update docs/names

eric-wieser commented on src/data/nat/choose/basic.lean:
```diff
@@ -169,6 +169,24 @@ begin
   exact mul_le_mul_left _ hk,
 end
 
+lemma desc_fac_eq_choose_mul_factorial (n k : ℕ) : desc_fac n k = k! * (n + k).choose k :=
```
Name and statement don't match
```suggestion
lemma desc_fac_eq_factorial_mul_choose (n k : ℕ) : desc_fac n k = k! * (n + k).choose k :=
```

eric-wieser commented on src/data/nat/choose/basic.lean:
```diff
@@ -169,6 +169,24 @@ begin
+  exact le_add_left k n
+end
+
+lemma fac_dvd_desc_fac (n k : ℕ) : k! ∣ desc_fac n k :=
```
```suggestion
lemma factorial_dvd_desc_fac (n k : ℕ) : k! ∣ desc_fac n k :=
```

eric-wieser commented on src/data/nat/factorial.lean:
```diff
@@ -143,4 +146,43 @@ begin
+lemma desc_fac_eq_div (n k : ℕ) : desc_fac n k = (n + k)! / n! :=
+begin
+  apply mul_left_cancel' (factorial_ne_zero n), rw factorial_mul_desc_fac,
+  exact (nat.mul_div_cancel' $ factorial_dvd_factorial $ le.intro rfl).symm
```
```suggestion
  apply mul_left_cancel' (factorial_ne_zero n),
  rw factorial_mul_desc_fac,
  exact (nat.mul_div_cancel' $ factorial_dvd_factorial $ le.intro rfl).symm
```

ericrbg commented on src/data/nat/choose/basic.lean:
```diff
@@ -169,6 +169,24 @@ begin
+begin
+  apply mul_left_cancel' (factorial_ne_zero k),
+  rw ←desc_fac_eq_choose_mul_factorial,
+  exact (nat.mul_div_cancel' $ fac_dvd_desc_fac _ _).symm
```
```suggestion
  exact (nat.mul_div_cancel' $ factorial_dvd_desc_fac _ _).symm
```

ericrbg committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
turns out I use lemmas I write

ericrbg committed:
:(

benjamindavidson commented on src/ring_theory/polynomial/bernstein.lean:
```diff
@@ -207,19 +182,19 @@ begin
     { simp [eval_at_0], },
-    { simp only [nat.succ_eq_add_one] at h,
+    { rw [nat.succ_eq_add_one] at h,
       have h' : ν ≤ n-1 := nat.le_sub_right_of_add_le h,
```
You can delete this line entirely.
```suggestion
    { have h' : ν ≤ n-1 := nat.le_sub_right_of_add_le h,
```

benjamindavidson commented on src/ring_theory/polynomial/bernstein.lean:
```diff
@@ -207,19 +182,19 @@ begin
+      { rw [show n - 1 - (ν - 1) = n - ν, by omega, pochhammer_eval_succ],
+      norm_cast,
+      rw show n - ν + ν = n, by omega }
+    } },
```
```suggestion
      by_cases h'' : ν = 0,
      { simp [h''] },
      { rw [show n - 1 - (ν - 1) = n - ν, by omega, pochhammer_eval_succ],
        rw_mod_cast show n - ν + ν = n, by omega } } },
```

ericrbg replied:
`rw_mod_cast`; that's very cool stuff! I'm gonna try it on some of the `pochhammer` rw too

benjamindavidson reviewed (commented):
A couple of minor comments, otherwise LGTM.

ericrbg committed:
use `rw_mod_cast`

ericrbg committed:
more rw_mod_cast and stop `rw`ing away `succ`

I'm in a bad habit with that because of `ring` not dealing with `succ` well...

Co-authored-by: Benjamin Davidson <68528197+benjamindavidson@users.noreply.github.com>

eric-wieser commented on src/ring_theory/polynomial/bernstein.lean:
```diff
@@ -207,19 +182,16 @@ begin
-      { have w₂ : n - 1 - (ν - 1) = n - ν, { rw [nat.sub_sub], rw nat.add_sub_cancel', omega, },
-        simpa [w₂] using (iterate_derivative_at_0_aux₂ R ν n), }, }, },
+      { simp [h''] },
+      { rw [show n - 1 - (ν - 1) = n - ν, by omega, pochhammer_eval_succ],
```
I think hiding the `omega` in the `rw` makes this harder to read:
```suggestion
      { have : n - 1 - (ν - 1) = n - ν := by omega,
        rw [this, pochhammer_eval_succ],
```

eric-wieser commented on src/ring_theory/polynomial/pochhammer.lean:
```diff
@@ -123,49 +132,29 @@ section factorial
+  exact or.inr rfl,
 end
+| (k + 1) := by rw_mod_cast [pochhammer_nat_eq_desc_fac, pochhammer_nat_eq_desc_fac,
+                             nat.succ_desc_fac, add_right_comm]
```
I think this proof works better as two lemmas:
```suggestion
lemma pochhammer_nat_eval_succ (r : ℕ) :
  ∀ n : ℕ, n * (pochhammer ℕ r).eval (n + 1) = (n + r) * (pochhammer ℕ r).eval n
| 0 := begin
  by_cases h : r = 0,
  { simp only [h, zero_mul, zero_add], },
  { simp only [pochhammer_eval_zero, zero_mul, if_neg h, mul_zero], }
end
| (k + 1) := by simp only [pochhammer_nat_eq_desc_fac, nat.succ_desc_fac, add_right_comm]

lemma pochhammer_eval_succ (r n : ℕ) :
  (n : S) * (pochhammer S r).eval (n + 1 : S) = (n + r) * (pochhammer S r).eval n :=
by exact_mod_cast congr_arg nat.cast (pochhammer_nat_eval_succ r n)

ericrbg replied:
My idea was that all `(pochhammer ℕ).eval` stuff should be done in `desc_fac`-land; but I guess this makes some sense as there's a slightly bigger domain over in `pochhammer` than `desc_fac`

eric-wieser replied:
That's a nice idea, but I'd argue your current proof doesn't achieve it. The aim in my suggestion is to push all the cast nonsense into a separate proof. Additionally the `pochhammer_eval_succ` lemma is hard to use when `S` is actually `nat`, because it's full of casts from `nat` to `nat` that have to first be removed. Adding `pochhammer_nat_eval_succ ` solves that problem, as it doesn't have any of the casts in the first place.

ericrbg replied:
yeah fair, I just messed with it a couple minutes and it ended up being a lot more of a pain than I expected to use it

ericrbg commented on src/ring_theory/polynomial/bernstein.lean:
```diff
@@ -207,19 +182,16 @@ begin
-        simpa [w₂] using (iterate_derivative_at_0_aux₂ R ν n), }, }, },
+      { simp [h''] },
+      { rw [show n - 1 - (ν - 1) = n - ν, by omega, pochhammer_eval_succ],
+        rw_mod_cast show n - ν + ν = n, by omega } } },
```
```suggestion
        have : n - ν + ν = n := by omega,
        rw_mod_cast this } } },
```

ericrbg committed:
improve readability

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ericrbg committed:
split `eval_succ`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
